ui: Bin order tables by rate.

martonp opened the pull request:
This PR updates the order tables on the market UI. Instead of displaying each individual order's quantity and rate on a separate row, the total quantity available at each rate will be displayed. Also, as discussed in matrix, epoch orders will be binned along with all other orders, and the check mark will no longer be displayed.

Closes #945.

JoeGruffins reviewed (approved):
Looks perfect to me.

chappjc commented:
Appears to work as indented.

However, we should be aware that while this is a UX improvement to some, it actually hides some important information that I hope we can strive to reveal again, perhaps in a future PR.

Take the following simnet buy side entirely comprising my own orders:

![image](https://user-images.githubusercontent.com/9373513/121262223-6241a780-c879-11eb-94ba-d73033a3fbd2.png)

Another party just sees the Buy Orders table and has no idea if the 20 DCR best bid is one order or two.  This may be an important factor in the user deciding if the transaction will be maximally efficient if they decide to sell.  Even if the stars align and those maker orders do get batched by the taker/seller's redeem txn, that redeem is still larger than it would have been if there were just one match instead of two.

Then consider the 30 DCR at 0.003945.  If a user's sell fills 20 DCR at that price level, what would they match against? They would need to see the unbinned orders sorted by time stamp (as before) to know what to expect (assuming no other takers got to those bids first).

martonp commented:
@chappjc There could be a little pop up that shows up when you hover over each row that contains this information. I may be wrong, but I don’t think people would be so concerned with this if the lot size is set properly and the transaction fees are negligible compared to the order. It definitely can’t hurt to have it there though.

buck54321 commented on client/webserver/site/src/js/markets.js:
```diff
@@ -1464,23 +1482,28 @@ export default class MarketsPage extends BasePage {
-    if (row && row.order.rate === 0) row = row.nextSibling
-    const tr = this.orderTableRow(order, cssClass)
+    if (row && row.getRate() === 0) row = row.nextSibling
     while (row) {
-      if ((order.rate < row.order.rate) === order.sell) {
+      if (order.rate === row.getRate()) {
+        row.insertOrder(order)
+        return
+      } else if ((order.rate < row.getRate()) === order.sell) {
+        const tr = this.orderTableRow([order], cssClass)
         tbody.insertBefore(tr, row)
         return
       }
       row = row.nextSibling
     }
```
You'll need to figure out a way to separate epoch queue orders and booked orders and create a row for each.

buck54321 commented on client/webserver/site/src/js/markets.js:
```diff
@@ -1545,33 +1567,68 @@ export default class MarketsPage extends BasePage {
+      const index = tr.orderBin.findIndex(order => order.token === token)
+      if (index < 0) return false
+      tr.orderBin.splice(index, 1)
+      if (!tr.orderBin.length) {
+        tr.remove()
+      } else {
+        tr.updateQty()
+      }
```
Can be two lines

buck54321 commented on client/webserver/site/src/js/markets.js:
```diff
@@ -1545,33 +1567,68 @@ export default class MarketsPage extends BasePage {
+      tr.orderBin = tr.orderBin.filter((order) => {
+        return !(order.epoch && order.epoch !== newEpoch)
+      })
+      if (!tr.orderBin.length) {
+        tr.remove()
+      } else {
+        tr.updateQty()
+      }
```
Can be two lines.

buck54321 reviewed (changes requested):
Looking good for a first pass. I do want to separate epoch orders from booked orders.

As far as the UI, I think a good first approach would just be to label the quantity of orders in the bin. e.g.

![image](https://user-images.githubusercontent.com/6109680/122226105-404cb400-ce7b-11eb-8f6e-7800dbc74f9c.png)

You could do that in a separate PR too.

martonp commented:
@buck54321 I've added a new column that shows the number of orders, let me know what you think. Also the epoch and non-epoch orders are now in separate rows.

martonp commented on client/webserver/site/package.json:
```diff
@@ -23,7 +23,7 @@
     "clean-webpack-plugin": "^3.0.0",
     "css-loader": "^5.1.2",
-    "css-minimizer-webpack-plugin": "^1.2.0",
+    "css-minimizer-webpack-plugin": "^3.0.1",
```
The CI check failed due to a vulnerability found in this package. It is now passing again after updating this.

chappjc replied:
I don't think CI failed because of a vuln in this. https://github.com/decred/dcrdex/runs/2869810692#step:4:6
Appears to be some transient npm error.

Rather than have a  +7,140 −10,057 diff, pls revers this last diff and we'll address it in another PR.
I have a branch with a batch of npm dep updates.

chappjc commented:
> @buck54321 I've added a new column that shows the number of orders, let me know what you think. Also the epoch and non-epoch orders are now in separate rows.

I think it uses a bit too much space.  Just a number and no bulky column header like @buck54321 showed would be fine IMO.  Here's what I see now:

![image](https://user-images.githubusercontent.com/9373513/124617154-53302400-de3c-11eb-9523-ce8c128684f2.png)

But it gets a little glitchy at various window widths, putting "# Orders" on two rows, and often creating a horizontal scroll bar:

![image](https://user-images.githubusercontent.com/9373513/124617360-7e1a7800-de3c-11eb-939b-03d417030f58.png)

Otherwise, it works as intended.  There's a separate row for epoch orders, and when booked, it merges that with the booked orders row and updates the # Orders count.  No issues in browser JS console.  Overall nice work, but I don't think the column header adds much, possibly even the separate column is unnecessary if the number is just after the quantity in the same column with appropriate styling.  Like buck suggested, but perhaps with a tweaked styling:

![image](https://user-images.githubusercontent.com/9373513/124617900-ee28fe00-de3c-11eb-91a6-d8d9f51921f0.png)

chappjc reviewed (changes requested):
> @buck54321 I've added a new column that shows the number of orders, let me know what you think. Also the epoch and non-epoch orders are now in separate rows.

I think it uses a bit too much space.  Just a number and no bulky column header like @buck54321 showed would be fine IMO.  Here's what I see now:

![image](https://user-images.githubusercontent.com/9373513/124617154-53302400-de3c-11eb-9523-ce8c128684f2.png)

But it gets a little glitchy at various window widths, putting "# Orders" on two rows, and often creating a horizontal scroll bar:

![image](https://user-images.githubusercontent.com/9373513/124617360-7e1a7800-de3c-11eb-939b-03d417030f58.png)

Otherwise, it works as intended.  There's a separate row for epoch orders, and when booked, it merges that with the booked orders row and updates the # Orders count.  No issues in browser JS console.  Overall nice work, but I don't think the column header adds much, possibly even the separate column is unnecessary if the number is just after the quantity in the same column with appropriate styling.  Like buck suggested, but perhaps with a tweaked styling:

![image](https://user-images.githubusercontent.com/9373513/124617900-ee28fe00-de3c-11eb-91a6-d8d9f51921f0.png)

martonp commented:
@chappjc I've updated it. This definitely looks better. I thought it would be confusing to just have a number there without a column header, but I added a tooltip which explains what the number means.

chappjc reviewed (approved):
I like it.  Nice work!

![image](https://user-images.githubusercontent.com/9373513/124671036-7974b480-de7a-11eb-8c46-695bba0aa5e1.png)

@buck54321 LGTM, some frontend changes since your last review though.

buck54321 commented on client/webserver/site/src/js/markets.js:
```diff
@@ -1525,53 +1557,103 @@ export default class MarketsPage extends BasePage {
+      const mainDiv = qtyTD.children[0]
+      const [qtyDiv, ordersDiv] = mainDiv.children
+      const qty = tr.orderBin.reduce((total, curr) => total + curr.qty, 0)
+      qtyDiv.innerText = qty.toFixed(8)
+      ordersDiv.innerText = tr.orderBin.length
```
Let's only show the `qtyDiv` when there is more than one order.

buck54321 commented on client/webserver/site/src/css/market.scss:
```diff
@@ -427,6 +427,12 @@ table.balance-table button:hover {
 }
 
+div.numorders {
+  color: white;
```
No white text, please. Inherited `color` will work fine once you get the background colors adjusted.

buck54321 commented on client/webserver/site/src/css/market_dark.scss:
```diff
@@ -24,6 +24,11 @@ body.dark {
   }
 
+  div.numorders {
+    background-color: rgb(25, 25, 180);
```
Let's darken this up quite a bit. Maybe like a `#141488`.

chappjc replied:
Maybe we could test out one of our primary colors: https://decred.org/brand/

Keep the hue / rgb ratio and tweak the brightness?  If that clashes with the rest of the DEX scheme, no biggie.

martonp replied:
![Screen Shot 2021-07-08 at 12 28 08 PM](https://user-images.githubusercontent.com/6186350/124958555-4819f700-dfe8-11eb-84bc-8f2e7564c098.png)
![Screen Shot 2021-07-08 at 12 28 21 PM](https://user-images.githubusercontent.com/6186350/124958563-4c461480-dfe8-11eb-920f-39bb9379a181.png)

The first one is a lighter version of the dark blue primary color and the second is @buck54321 's suggestion. Very similar, but I think @buck54321 's suggestion is a closer match to the dex logo, which doesn't seem to be using any of the Decred colors.

chappjc replied:
k, no sweat.  Just wanted to try to keep us on-brand if possible, but best to stay internally consistent with other dex styles unless we rework the whole scheme.

buck54321 commented on client/webserver/site/src/css/market.scss:
```diff
@@ -427,6 +427,12 @@ table.balance-table button:hover {
+div.numorders {
+  color: white;
+  background-color: rgb(94, 94, 130);
+  border-radius: 3px;
+}
```
The coloring is a little bright. It should blend in and match existing styles more closely.

![image](https://user-images.githubusercontent.com/6109680/124758565-4ff86f00-def4-11eb-9d4a-b0ba9a3a4d0c.png)

To do that, I added `line-height: 1; padding: 1px;` here, and then removed the `px-1` class and added the `align-items-center` class to the wrapper `div`. You might want to play with `font-size: 0.9em` too, but the font-size is already pretty small here so I'm not sure we want to push it.

buck54321 commented on client/webserver/site/src/css/market.scss:
```diff
@@ -427,6 +427,12 @@ table.balance-table button:hover {
 
+div.numorders {
+  color: white;
+  background-color: rgb(94, 94, 130);
```
Let's lighten this up quite a bit. Maybe like a `#c5c5e6`

buck54321 commented on client/webserver/site/src/js/markets.js:
```diff
@@ -1525,53 +1557,103 @@ export default class MarketsPage extends BasePage {
-    tr.addQty = (qty) => {
-      tr.qty += qty
-      qtyTD.innerText = tr.qty.toFixed(8)
+    tr.insertOrder = (order) => {
+      tr.orderBin.push(order)
+      updateQtyTD()
+    }
+    tr.updateOrderQty = (token, qty) => {
+      for (let i = 0; i < tr.orderBin.length; i++) {
+        if (tr.orderBin[i].token === token) {
+          tr.orderBin[i].qty = qty
+          updateQtyTD()
+          return true
+        }
+      }
+      return false
+    }
+    tr.removeOrder = (token) => {
+      const index = tr.orderBin.findIndex(order => order.token === token)
+      if (index < 0) return false
+      tr.orderBin.splice(index, 1)
+      if (!tr.orderBin.length) tr.remove()
+      else updateQtyTD()
+      return true
+    }
+    tr.removeEpochOrders = (newEpoch) => {
+      tr.orderBin = tr.orderBin.filter((order) => {
+        return !(order.epoch && order.epoch !== newEpoch)
+      })
+      if (!tr.orderBin.length) tr.remove()
+      else updateQtyTD()
+    }
+    tr.getRate = () => {
+      return tr.orderBin[0].rate
+    }
+    tr.getEpoch = () => {
+      return tr.orderBin[0].epoch
+    }
+    tr.compare = (order) => {
```
This is good stuff, and it makes the semantics of updating these rows really clean, but once we're adding seven new functions as custom properties to an `td` element, that's a big clue to me that we should just write a custom class to manage the element and make these class methods.

martonp replied:
I've added a manager class that has a reference to the `td` element the `td` element has a reference to the manager class. It's definitely a lot cleaner.

buck54321 commented on client/webserver/site/src/js/markets.js:
```diff
@@ -1525,53 +1557,103 @@ export default class MarketsPage extends BasePage {
     })
     let qtyTD
+    const updateQtyTD = () => {
+      const mainDiv = qtyTD.children[0]
+      const [qtyDiv, ordersDiv] = mainDiv.children
```
Try using the `data-tmpl`/`Doc.tmplElement` system used throughout, and assigning these elements as properties of the custom class mentioned in my other comment.

buck54321 reviewed (changes requested):
Good stuff. I've made a few more suggestions to round things out.

chappjc reviewed (approved):
Re-approving latest changes.

buck54321 commented on client/webserver/site/src/js/markets.js:
```diff
@@ -1986,3 +1995,102 @@ function cleanTemplates (...tmpls) {
+    return this.sell
+  }
+
+  compare (order) {
```
I try to doc all methods, but this method in particular could use a docstring.

buck54321 commented on client/webserver/site/src/js/markets.js:
```diff
@@ -1446,14 +1446,43 @@ export default class MarketsPage extends BasePage {
   loadTableSide (sell) {
     const bookSide = sell ? this.book.sells : this.book.buys
     const tbody = sell ? this.page.sellRows : this.page.buyRows
     const cssClass = sell ? 'sellcolor' : 'buycolor'
```
`cssClass` is unused outside of `OrderTableRowManager` (via `orderTableRow`). I'd recommend moving this `cssClass` stuff into the `OrderTableRowManager` constructor, and just passing the `sell` boolean as an argument instead.

martonp replied:
Good call.

buck54321 reviewed (changes requested):
If I place two identical orders in quick succession, the epoch orders are not binned.

![image](https://user-images.githubusercontent.com/6109680/125122404-fdb57a80-e0ba-11eb-9943-eeda33168e01.png)

Then, when the epoch expires, and the orders are booked, they are properly grouped, but the epoch check mark is still displayed.

![image](https://user-images.githubusercontent.com/6109680/125122521-289fce80-e0bb-11eb-9406-748dd8923144.png)

martonp commented:
@buck54321 The bug is fixed now.
